src/Controls/src/Core/DispatcherExtensions.cs

```diff
@@ -28,28 +29,62 @@ public static IDispatcher GetDispatcher(this BindableObject? bindableObject)
 			throw new InvalidOperationException("BindableObject was not instantiated on a thread with a dispatcher nor does the current application have a dispatcher.");
 		}
 
-		public static void Dispatch(this IDispatcher? dispatcher, Action action)
+		public static void DispatchIfRequired(this IDispatcher? dispatcher, Action action)
 		{
-			// try looking on the app
-			if (dispatcher is null && Application.Current?.Dispatcher is IDispatcher appDispatcher)
-				dispatcher = appDispatcher;
-
-			// maybe this thread has a dispatcher
-			if (dispatcher is null)
-				dispatcher = Dispatcher.GetForCurrentThread();
+			dispatcher = EnsureDispatcher(dispatcher);
+			if (dispatcher.IsDispatchRequired)
+			{
+				dispatcher.Dispatch(action);
+			}
+			else
+			{
+				action();
+			}
+		}
 
-			// no dispatchers found at all
-			if (dispatcher is null)
-				throw new InvalidOperationException("The dispatcher was not found and the current application does not have a dispatcher.");
+		public static Task DispatchIfRequiredAsync(this IDispatcher? dispatcher, Action action)
+		{
+			dispatcher = EnsureDispatcher(dispatcher);
+			if (dispatcher.IsDispatchRequired)
+			{
+				return dispatcher.DispatchAsync(action);
+			}
+			else
+			{
+				action();
+				return Task.CompletedTask;
+			}
+		}
 
-			if (dispatcher.IsInvokeRequired)
+		public static Task DispatchIfRequiredAsync(this IDispatcher? dispatcher, Func<Task> action)
+		{
+			dispatcher = EnsureDispatcher(dispatcher);
+			if (dispatcher.IsDispatchRequired)
 			{
-				dispatcher.BeginInvokeOnMainThread(action);
+				return dispatcher.DispatchAsync(action);
 			}
 			else
 			{
 				action();
```

Comment:
`return action();` ? The task returned is not being awaited in this case. It is being awaited inside the [DispatchAsync](https://github.com/dotnet/maui/blob/94781e8a70bee327b76edb179d032d006b8327ea/src/Core/src/Dispatching/DispatcherExtensions.cs#L9-L27) though.

Reply:
Good catch!